offload/plugins-nextgen/amdgpu/src/rtl.cpp

```diff
@@ -4642,13 +4647,16 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
   Error preAllocateDeviceMemoryPool() {
 
     void *DevPtr;
+    // Use PER_DEVICE_PREALLOC_SIZE (128KB) as heap and allocate 512MB for
+    // device memory
+    size_t PreAllocSize = hsa_utils::PER_DEVICE_PREALLOC_SIZE + DMSlabSize;
+
     for (AMDGPUMemoryPoolTy *MemoryPool : AllMemoryPools) {
       if (!MemoryPool->isGlobal())
         continue;
 
       if (MemoryPool->isCoarseGrained()) {
         DevPtr = nullptr;
-        size_t PreAllocSize = hsa_utils::PER_DEVICE_PREALLOC_SIZE;
 
         Error Err = MemoryPool->allocate(PreAllocSize, &DevPtr);
```

Comment:
I'm a little concerned about doing this in a single allocation.  The slabs must be 2MB aligned.  With the prealloc at the beginning, it seems unlikely that the slabs will end up being 2MB aligned.  Putting it at the end might raise the possibility, or allocating it separately.

We might want an assert or some other guard to ensure that DMSlabPtr is 2MB aligned.